include/ipc/ipc_service.h

```diff
@@ -203,6 +239,145 @@ int ipc_service_register_endpoint(const struct device *instance,
  */
 int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len);
 
+/** @brief Get an empty TX buffer to be sent using @ref ipc_service_send_nocopy
+ *
+ *  This function can be called to get an empty TX buffer so that the
+ *  application can directly put its data into the sending buffer without copy
+ *  from an application buffer.
+ *
+ *  It is the application responsibility to correctly fill the allocated TX
+ *  buffer with data and passing correct parameters to @ref
+ *  ipc_service_send_nocopy function to perform data no-copy-send mechanism.
+ *
+ *  The size parameter can be used to request a buffer with a certain size. If
+ *  the size can be accomodated the function returns no errors and the size
+ *  value is unchanged. If the requested size is too big, the function returns
+ *  -ENOMEM and the size parameter must contain the maximum allowed size.
```

Comment:
I assume the buffer is allocated only when this function returns 0. When it returns other error codes, the buffer is not allocated and the caller is not responsible to drop it. I think this function doc should clearly describe if my assumption is correct.

Reply:
So we have two possibilities here:
1. If the size cannot be accomodate the function returns `-ENOMEM` and no buffer is allocated so the caller is not responsible to drop it. This could be a problem with OpenAMP because AFAIK there is no way no determine the size without allocation the buffer (***)
2. The function returns `-ENOMEM` but the buffer is allocated anyway and the maximum size is still returned to the caller using the `size` parameter. This is the solution implemented in the static vrings backend. https://github.com/zephyrproject-rtos/zephyr/blob/89ef492572edfccb6efcabd28ba7e2f6a8297aac/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L497-L507

*** @arnopo is there an OpenAMP API to get the `len` returned by `rpmsg_get_tx_payload_buffer()` without actually allocating the buffer?

Reply:
I think this API does not have to be fully aligned with OpenAMP. If OpenAMP has this limitation, implementation here could call `rpmsg_free_tx_buffer()` (coming soon) if we decide to not allocate an buffer in `-NOMEM` scenario.

Reply:
I'm still unclear which way we want to go. Whether we want to still allocate and return the buffer or not.

I think an even better solution could be make the `len` parameter a `desirable` buffer size. If the backend could accomodate that ok, in all the other cases (the backend is able to allocate either a smaller or a bigger buffer) the function should return `0` (because the buffer was allocated) and the allocated size in `*len`. The user can then decide if that's fine (and use the buffer with the returned size) or if he needs to drop it.

@hubertmis @nordic-krch what do you think?

Reply:
As a user of this API I would like it to allocate the size I'm asking for. If implementation cannot give me what I'm asking, I would expect an error like `-ENOMEM`, possibly accompanied with an information what smaller buffer size could have been allocated.
As API user most probably I'm not interested in this smaller buffer size, because the data I need to send would not fit. Because of this I would prefer to receive `-ENOMEM` and not get any buffers allocated.
If for any reason I could use smaller buffer, I can use the accompanying information to request again.

Reply:
ok fair enough.